packages/cli-kit/src/session/exchange.ts

```diff
@@ -152,42 +159,47 @@ async function requestAppToken(
   if (api === 'admin' && store) {
     identifier = `${store}-${appId}`
   }
-  const appToken = await tokenRequest(params).then(buildApplicationToken)
+  const tokenResult = await tokenRequest(params)
+  const value = tokenResult.mapError(tokenRequestErrorHandler).valueOrThrow()
+  const appToken = await buildApplicationToken(value)
   return {[identifier]: appToken}
 }
 
-// eslint-disable-next-line @typescript-eslint/no-explicit-any
-async function tokenRequest(params: {[key: string]: string}, throws = true): Promise<any> {
+interface TokenRequestResult {
+  access_token: string
+  expires_in: number
+  refresh_token: string
+  scope: string
+}
+
+function tokenRequestErrorHandler(error: string) {
+  if (error === 'invalid_grant') {
+    // There's an scenario when Identity returns "invalid_grant" when trying to refresh the token
+    // using a valid refresh token. When that happens, we take the user through the authentication flow.
+    throw new InvalidGrantError()
+  }
+  if (error === 'invalid_request') {
+    // There's an scenario when Identity returns "invalid_request" when exchanging an identity token.
+    // This means the token is invalid. We clear the session and throw an error to let the caller know.
+    throw InvalidIdentityError
```

Comment:
I'd change this one from:

```js
const InvalidIdentityError = new Abort(
  '\nError validating auth session',
  "We've cleared the current session, please try again",
)
```

To: 

```js
const InvalidIdentityError = () => new Abort(
  '\nError validating auth session',
  "We've cleared the current session, please try again",
)
```
Otherwise the stacktrace won't be accurate because it's captured when the error is instantiated. Then this piece of code should be:

```js
throw InvalidIdentityError()
```

Reply:
good catch 👌